File: dodo/dodo/boj12100.cpp
#include<iostream>
#include<vector>
#include<algorithm>
#include<deque>
using namespace std;

int N;
int ans = 0;
deque<int> dq;
vector<vector<int>> move_left(vector<vector<int>> m) {
	for (int i = 0; i < N; i++) {
		for (int j = 0; j < N; j++)
			if (m[i][j] != 0) {
				dq.push_back(m[i][j]);
				m[i][j] = 0;
			}
		int idx = 0;
		while (!dq.empty()) {
			int cur = dq.front();
			dq.pop_front();
			if (dq.size() >= 1) {
				if (dq.front() == cur) {
					m[i][idx++] = cur * 2;
					dq.pop_front();
				}
				else
					m[i][idx++] = cur;
			}
			else
				m[i][idx++] = cur;
		}
	}
	return m;
}
vector<vector<int>> move_right(vector<vector<int>> m) {

	for (int i = 0; i < N; i++) {
		for (int j = N - 1; j >= 0; j--)
			if (m[i][j] != 0) {
				dq.push_back(m[i][j]);
				m[i][j] = 0;
			}
		int idx = N - 1;
		while (!dq.empty()) {
			int cur = dq.front();
			dq.pop_front();
			if (dq.size() >= 1) {
				if (dq.front() == cur) {
					m[i][idx--] = cur * 2;
					dq.pop_front();
				}
				else
					m[i][idx--] = cur;
			}
			else
				m[i][idx--] = cur;
		}
	}
	return m;
}
vector<vector<int>> move_up(vector<vector<int>> m) {
	for (int i = 0; i < N; i++) {
		for (int j = 0; j < N; j++)
			if (m[j][i] != 0) {
				dq.push_back(m[j][i]);
				m[j][i] = 0;
			}
		int idx = 0;
		while (!dq.empty()) {
			int cur = dq.front();
			dq.pop_front();
			if (dq.size() >= 1) {
				if (dq.front() == cur) {
					m[idx++][i] = cur * 2;
					dq.pop_front();
				}
				else
					m[idx++][i] = cur;
			}
			else
				m[idx++][i] = cur;
		}
	}
	return m;
}
vector<vector<int>> move_down(vector<vector<int>> m) {
	for (int i = 0; i < N; i++) {
		for (int j = N - 1; j >= 0; j--)
			if (m[j][i] != 0) {
				dq.push_back(m[j][i]);
				m[j][i] = 0;
			}
		int idx = N - 1;
		while (!dq.empty()) {
			int cur = dq.front();
			dq.pop_front();
			if (dq.size() >= 1) {
				if (dq.front() == cur) {
					m[idx--][i] = cur * 2;
					dq.pop_front();
				}
				else
					m[idx--][i] = cur;
			}
			else
				m[idx--][i] = cur;
		}
	}
	return m;
}

void solve(int cnt, vector<vector<int>> m) {
	if (cnt == 5) {
		for (int i = 0; i < N; i++) 
			for (int j = 0; j < N; j++) 
				ans = max(ans, m[i][j]);

		return;
	}
	vector<vector<int>> temp;
	temp = move_left(m);
	solve(cnt + 1, temp);
	temp = move_right(m);
	solve(cnt + 1, temp);
	temp = move_up(m);
	solve(cnt + 1, temp);
	temp = move_down(m);
	solve(cnt + 1, temp);
}
int main() {
	cin >> N;
	vector<vector<int>> map(N, vector<int>(N, 0));
	for (int i = 0; i < N; i++) {
		for (int j = 0; j < N; j++) {
			cin >> map[i][j];
		}
	}
	solve(0, map);
	cout << ans << '\n';
	return 0;
}

File: doja/src/doja/BOJ12100.java
package doja;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Deque;
import java.util.LinkedList;
import java.util.StringTokenizer;

public class BOJ12100 {

	static int N;
	static int[][] map;
	static int[][] temp;
	static int ans=0;
	static void solve(int cnt,int val, int[][] m) {
		if(cnt==5) {
			ans=Math.max(val, ans);
			return;
		}
		Deque<Integer> dq = new LinkedList<>();
		temp = new int[N + 1][N + 1];
		//left
		int maxVal=0;
		for (int i = 0; i < N; i++) {
			// 수를 모두 덱에넣기
			for (int j = 0; j < N; j++) {
				if(m[i][j]==0)
					continue;
				dq.addLast(m[i][j]);
			}
			int idx = 0;
			while (!dq.isEmpty()) {
				int cur = dq.pollFirst();
				if (!dq.isEmpty()) {
					if (cur == dq.peekFirst()) {
						dq.pollFirst();
						maxVal=Math.max(maxVal, cur*2);
						temp[i][idx++] = cur * 2;
					} else {
						maxVal=Math.max(maxVal, cur);
						temp[i][idx++]=cur;
					}
				}
				else {
					maxVal=Math.max(maxVal, cur);
					temp[i][idx++]=cur;
				};
			}
		}
		solve(cnt+1,maxVal,temp);
		//right
		temp = new int[N + 1][N + 1];
		maxVal=0;
		for (int i = 0; i < N; i++) {
			// 수를 모두 덱에넣기
			for (int j = N-1; j >=0 ; j--) {
				if(m[i][j]==0)
					continue;
				dq.addLast(m[i][j]);
			}
			int idx = N-1;
			while (!dq.isEmpty()) {
				int cur = dq.pollFirst();
				if (!dq.isEmpty()) {
					if (cur == dq.peekFirst()) {
						dq.pollFirst();
						maxVal=Math.max(maxVal, cur*2);
						temp[i][idx--] = cur * 2;
					} else {
						maxVal=Math.max(maxVal, cur);
						temp[i][idx--]=cur;
					}
				}
				else {
					maxVal=Math.max(maxVal, cur);
					temp[i][idx--]=cur;
				}
			}
		}
		solve(cnt+1,maxVal,temp);
		//up
		temp = new int[N + 1][N + 1];
		maxVal=0;
		for (int i = 0; i < N; i++) {
			// 수를 모두 덱에넣기
			for (int j = 0; j <N ; j++) {
				if(m[j][i]==0)
					continue;
				dq.addLast(m[j][i]);
			}
			int idx = 0;
			while (!dq.isEmpty()) {
				int cur = dq.pollFirst();
				if (!dq.isEmpty()) {
					if (cur == dq.peekFirst()) {
						dq.pollFirst();
						maxVal=Math.max(maxVal, cur*2);
						temp[idx++][i] = cur * 2;
					} else {
						maxVal=Math.max(maxVal, cur);
						temp[idx++][i]=cur;
					}
				}
				else {
					maxVal=Math.max(maxVal, cur);
					temp[idx++][i]=cur;
				}
			}
		}
		solve(cnt+1,maxVal,temp);
		//down
		temp = new int[N + 1][N + 1];
		maxVal=0;
		for (int i = 0; i < N; i++) {
			// 수를 모두 덱에넣기
			for (int j = N-1; j >=0 ; j--) {
				if(m[j][i]==0)
					continue;
				dq.addLast(m[j][i]);
			}
			int idx = N-1;
			while (!dq.isEmpty()) {
				int cur = dq.pollFirst();
				if (!dq.isEmpty()) {
					if (cur == dq.peekFirst()) {
						dq.pollFirst();
						maxVal=Math.max(maxVal, cur*2);
						temp[idx--][i] = cur * 2;
					} else {
						maxVal=Math.max(maxVal, cur);
						temp[idx--][i]=cur;
					}
				}
				else {
					maxVal=Math.max(maxVal, cur);
					temp[idx--][i]=cur;
				}
			}
		}
		solve(cnt+1,maxVal,temp);
	}

	public static void main(String[] args) throws IOException {
		// TODO Auto-generated method stub
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		N = Integer.parseInt(br.readLine());
		map = new int[N + 1][N + 1];
		for (int i = 0; i < N; i++) {
			StringTokenizer tok = new StringTokenizer(br.readLine());
			for (int j = 0; j < N; j++) {
				map[i][j] = Integer.parseInt(tok.nextToken());
			}
		}
		solve(0,0,map);
		System.out.println(ans);
	}

}
